src/Parser/Mage/Frost/Modules/Features/BrainFreeze.js

```diff
@@ -11,81 +11,102 @@ const debug = false;
 // Brain Freeze appears to always fall after Flurry cast, but not always on same timestamp. Giving a margin here.
 const PROC_WINDOW_MS = 100;
 
+// When Glacial Spike is talented, it's OK to overwrite Brain Freeze procs if you have at least this many icicles during the frost bolt cast
+const ICICLE_BREAKPOINT = 3;
+
 class BrainFreeze extends Analyzer {
-	lastFlurryTimestamp;
+  lastFlurryTimestamp;
 
-	overwrittenProcs = 0;
-	expiredProcs = 0;
-	totalProcs = 0;
-	flurryWithoutProc = 0;
+  overwrittenProcs = 0;
+  okOverwrittenProcs = 0;
+  expiredProcs = 0;
+  totalProcs = 0;
+  flurryWithoutProc = 0;
 
-	on_byPlayer_applybuff(event) {
-		const spellId = event.ability.guid;
-    if (spellId !== SPELLS.BRAIN_FREEZE.id) {
-			return;
-		}
-		this.totalProcs += 1;
-	}
+  constructor(...args) {
+    super(...args);
+    this.glacialSpikeTalented = this.selectedCombatant.hasTalent(SPELLS.GLACIAL_SPIKE_TALENT.id);
+  }
+
+  on_byPlayer_applybuff(event) {
+    const spellId = event.ability.guid;
+    if (spellId === SPELLS.BRAIN_FREEZE.id) {
+      this.totalProcs += 1;
+    }
+  }
 
-	on_byPlayer_refreshbuff(event) {
-		const spellId = event.ability.guid;
+  on_byPlayer_refreshbuff(event) {
+    const spellId = event.ability.guid;
     if (spellId !== SPELLS.BRAIN_FREEZE.id) {
-			return;
-		}
-		this.overwrittenProcs += 1;
-		this.totalProcs += 1;
-    debug && console.log("Brain Freeze proc overwritten @ " + formatMilliseconds(event.timestamp - this.owner.fight.start_time));
+      return;
+    }
+    this.totalProcs += 1;
+
+    if (!this.glacialSpikeTalented) {
+      this.overwrittenProcs += 1;
+      debug && console.log("Brain Freeze proc overwritten w/o GS talented @ " + formatMilliseconds(event.timestamp - this.owner.fight.start_time));
+      return;
+    }
+
+    // Get the number of icicles the player had during the cast of the frostbolt
+    const stacksDuringCast = this.selectedCombatant.getBuff(SPELLS.ICICLES_BUFF.id, this.owner.currentTimestamp - 250).stacks || 0;
```

Comment:
So one issue with his this works and how the events seem to stack up.

Lets say the player has 3 Icicles when they begin casting in each of these scenarios.

1. The player starts casting Ebonbolt. Ebonbolt doesnt give icicles, so getBuff is going to say that they had 3 Icicles when they gained the Brain Freeze Proc, and that is correct.

2. The player starts casting Frostbolt. When Frostbolt completes, the event order is as follows
![image](https://user-images.githubusercontent.com/1304554/45602243-410d3e80-b9e0-11e8-98fa-bef4878d760f.png)

So in this case, getBuff would also say that they had 3 Icicles, but they had actually just gained the 4th.

In Scenario 1, they should be using the Brain Freeze cause they have less than 4 Icicles whereas in Scenario 2 they should be holding it because they would be gaining their 4th Icicle from Frostbolt.